module/bdev/lvol/vbdev_lvol.c

```diff
@@ -2087,4 +2089,228 @@ vbdev_lvol_shallow_copy(struct spdk_lvol *lvol, const char *bdev_name,
 	spdk_lvol_shallow_copy(lvol, ext_dev, _vbdev_lvol_shallow_copy_cb, req);
 }
 
+struct fragmap_io {
+	struct spdk_bdev *bdev;
+	struct spdk_bdev_desc *bdev_desc;
+	struct spdk_io_channel *bdev_io_channel;
+	struct spdk_jsonrpc_request *request;
+
+	struct spdk_bit_array *fragmap;
+
+	uint64_t cluster_size;
+	uint64_t block_size;
+	uint64_t num_clusters;
+	uint64_t num_allocated_clusters;
+
+	uint64_t offset;
+	uint64_t size;
+	uint64_t current_offset;
+};
+
+
+static void seek_hole_done_cb(struct spdk_bdev_io *bdev_io, bool success, void *cb_arg);
+
+static void
+get_fragmap_done(struct fragmap_io *io, int error_code, const char *error_msg)
+{
+	struct spdk_json_write_ctx *w = NULL;
+	char *encoded;
+
+	if (error_code != 0) {
+		spdk_jsonrpc_send_error_response_fmt(io->request, error_code, "%s: %s",
+						     error_msg, spdk_strerror(-error_code));
+		goto cleanup;
+	} 
+
+	encoded = spdk_bit_array_to_base64_string(io->fragmap);
+	if (encoded == NULL) {
+		SPDK_ERRLOG("Failed to encode fragmap to base64 string\n");
+		spdk_jsonrpc_send_error_response_fmt(io->request, SPDK_JSONRPC_ERROR_INTERNAL_ERROR,
+						     "failed to encode fragmap");
+		goto cleanup;
+	}
+
+	w = spdk_jsonrpc_begin_result(io->request);
```

Comment:
For better separation, you should pass back the result of the operation to the rpc callback `rpc_bdev_lvol_get_fragmap_cb` and only there write the json response, like for example the rpc `rpc_bdev_lvol_snapshot` does:

- RPC layer handle the json request and response
- vbdev layer perform the lvol operation

Reply:
Addressed. Thank you.